src/libasr/codegen/asr_to_c_cpp.h

```diff
@@ -582,7 +582,7 @@ R"(#include <stdio.h>
         self().visit_expr(*x.m_arg);
         switch (x.m_kind) {
             case (ASR::cast_kindType::IntegerToReal) : {
-                src = "(float)(" + src + ")";
+                src = "(double)(" + src + ")";
```

Comment:
It should depend on the destination kind ideally. If the destination is 64-bit real then we should do `double` otherwise `float`. Doing `double` all the time might not be a good idea.

Reply:
Instead of this update the definition of `_lpython_floordiv` to use `f64` for doing and storing division results,

https://github.com/lcompilers/lpython/blob/0fc1847c56b470521a2afe1bef983facd6f6c220/src/runtime/lpython_builtin.py#L457

can be changed to `r = float(a)/float(b)` because `f32` rounds things up.

Reply:
Please, could you possibly share how `float()` could cast to a `double`? Currently, it seems `float(x)` casts to `f32` instead of the desired `f64`. 

I tried the suggested change. Please, find the relevant function from the output of `--show-c` below.  It casts the `a` and `b` to `f32`

```c
int32_t __lpython_overloaded_2___lpython_floordiv(int32_t a, int32_t b)
{
    int32_t _lpython_return_variable;
    double r;
    int32_t result;
    r = (float)(a)/(float)(b);
    result = (int)(r);
    if (r >=   0.00000000000000000e+00 || (float)(result) == r) {
        _lpython_return_variable = result;
        return _lpython_return_variable;
    }
    _lpython_return_variable = result - 1;
    return _lpython_return_variable;
}
```

It seem we need the line ` r = (float)(a)/(float)(b);` to be ` r = (double)(a)/(double)(b);`.

Reply:
> Please, could you possibly share how `float()` could cast to a `double`? Currently, it seems `float(x)` casts to `f32` instead of the desired `f64`.

I am talking about `float` intrinsic in Python. It always casts (or is supposed to cast to `f64`). Its different C/C++ `float`. The change I recommended in https://github.com/lcompilers/lpython/pull/1067#discussion_r959185801 should be done in the following functions,

https://github.com/lcompilers/lpython/blob/a227df8692cef722eb2630b64c0aa88b133512ed/src/runtime/lpython_builtin.py#L441-L472

C/C++ backend shouldn't be changed incorrectly to make `_lpython_floordiv` work.

Reply:
https://github.com/lcompilers/lpython/blob/a227df8692cef722eb2630b64c0aa88b133512ed/src/lpython/semantics/python_ast_to_asr.cpp#L1302-L1304

It seems that for division, the kind is already taken to be `8`. Please, possibly share if we still need to cast integers to floats using `float()` in the above functions that you shared.

Currently, it seems the `C` and `CPP` backends did not yet have support for casting `integers` to `double`. I attempted to add this support in https://github.com/lcompilers/lpython/pull/1067/commits/0613bd2f46c40eb6e647d121120376c024ec6d89, https://github.com/lcompilers/lpython/pull/1067/commits/ccc29fd3b1de0a9b40a0161059ec7456fe660ff3. Please possibly share if it is fine.

Reply:
> It seems that for division, the kind is already taken to be 8.

That's incorrect IMO. If both operands are 32-bit floating points then the result should also be 32-bit floating point. 

> Please, possibly share if we still need to cast integers to floats using float() in the above functions that you shared.

We should still do `float()` in Python code. Python code should be correct, its the compiler that needs to be fixed.

Reply:
Got it.

Reply:
I just added a commit to use `float()` to convert `i32` to `f64`. Gangandeep Sir, please, could you possibly share if we also need to cast `f32` to `f64` in `def _lpython_floordiv(a: f32, b: f32) -> f32`?

Reply:
Yes. Otherwise it will round things up. You can play with to find such case yourself. I figured one case while working on other PR.

Reply:
Got it.